src/common/config_parser.cpp

```diff
@@ -143,6 +143,15 @@ void ConfigParser::addOptionsGeneral(cli::CLIWrapper& cli) {
   cli.add<std::string>("--dump-config",
     "Dump current (modified) configuration to stdout and exit. Possible values: full, minimal, expand")
     ->implicit_val("full");
+  if(mode_ == cli::mode::training) {
+    // --sigterm is deliberately not a boolean, to allow for a consistent
+    // pattern of specifying custom signal handling in the future.
+    // (e.g., dump model but continue training upon SIGUSR1, or report current
+    // training status upon SIGINFO.)
+    cli.add<std::string>("--sigterm",
+      "What to do with SIGTERM: 'graceful' => save and exit (default); 'immediate' => exit immediately.", 
```

Comment:
"graceful" is not descriptive. Please say in the option string what it does, e.g. "save-and-exit"

"immediate" could be "terminate"

Reply:
Strings changed, but SIGTERM is widely considered a request for a graceful shutdown.
See here: https://qph.fs.quoracdn.net/main-qimg-1180ef2465c309928b02481f02580c6a
and here: https://www.booleanworld.com/kill-process-linux/#:~:text=By%20default%2C%20all%20the%20process,gracefully%22.
and here: https://github.com/python-trio/snekomatic/issues/6

Reply:
In all of these examples, graceful shutdown means to leave the world in a well-defined state (e.g. finish in-flight requests to a service). The examples do not suggest saving state, and note that the PR saves state in a way that is inconsistent with configured behavior, because the configuration says to save every `save-freq` updates, and not at a time of receiving SIGTERM, which is a random update.

In my mind, you don't want a "graceful shutdown" as given in these examples, but more like an "early exit" from the training process.

Alternatively, you may be wanting "preemption," i.e. save state that allows to restart the training. Your PR accomplishes that as well (right?), and such feature would be useful in our internal server farm as well when running jobs in low-priority preemptable mode.

Reply:
Saving state to resume later is what Microsoft incidentally calls 'graceful' preemption, as opposed to 'immediate' preemption: https://docs.microsoft.com/en-us/powershell/high-performance-computing/understanding-policy-configuration?view=hpc19-ps#:~:text=Immediate%20preemption%20(Default)%3A%20Take,allocated%20to%20another%20job%20immediately.&text=Graceful%20preemption%3A%20Take%20resources%20from,that%20work%20is%20not%20lost.